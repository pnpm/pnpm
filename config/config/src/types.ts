import npmTypes from '@pnpm/npm-conf/lib/types'

export const types = Object.assign({
  'auto-install-peers': Boolean,
  bail: Boolean,
  'cache-dir': String,
  'child-concurrency': Number,
  'merge-git-branch-lockfiles': Boolean,
  'merge-git-branch-lockfiles-branch-pattern': Array,
  color: ['always', 'auto', 'never'],
  'config-dir': String,
  'deploy-all-files': Boolean,
  'dedupe-peer-dependents': Boolean,
  'dedupe-direct-deps': Boolean,
  'dedupe-injected-deps': Boolean,
  dev: [null, true],
  dir: String,
  'disallow-workspace-cycles': Boolean,
  'enable-modules-dir': Boolean,
  'enable-pre-post-scripts': Boolean,
  'exclude-links-from-lockfile': Boolean,
  'extend-node-path': Boolean,
  'fetch-timeout': Number,
  'fetching-concurrency': Number,
  filter: [String, Array],
  'filter-prod': [String, Array],
  'frozen-lockfile': Boolean,
  'git-checks': Boolean,
  'git-shallow-hosts': Array,
  'global-bin-dir': String,
  'global-dir': String,
  'global-path': String,
  'global-pnpmfile': String,
  'git-branch-lockfile': Boolean,
  hoist: Boolean,
  'hoist-pattern': Array,
  'hoist-workspace-packages': Boolean,
  'ignore-compatibility-db': Boolean,
  'ignore-dep-scripts': Boolean,
  'ignore-pnpmfile': Boolean,
  'ignore-workspace': Boolean,
  'ignore-workspace-cycles': Boolean,
  'ignore-workspace-root-check': Boolean,
  'include-workspace-root': Boolean,
  'legacy-dir-filtering': Boolean,
  'link-workspace-packages': [Boolean, 'deep'],
  lockfile: Boolean,
  'lockfile-dir': String,
  'lockfile-directory': String, // TODO: deprecate
  'lockfile-include-tarball-url': Boolean,
  'lockfile-only': Boolean,
  loglevel: ['silent', 'error', 'warn', 'info', 'debug'],
  maxsockets: Number,
  'modules-cache-max-age': Number,
  'dlx-cache-max-age': Number,
  'modules-dir': String,
  'network-concurrency': Number,
  'node-linker': ['pnp', 'isolated', 'hoisted'],
  noproxy: String,
  'npm-path': String,
  offline: Boolean,
  'only-built-dependencies': [String],
  'pack-gzip-level': Number,
  'package-import-method': ['auto', 'hardlink', 'clone', 'copy'],
  'patches-dir': String,
  pnpmfile: String,
  'package-manager-strict': Boolean,
  'package-manager-strict-version': Boolean,
  'prefer-frozen-lockfile': Boolean,
  'prefer-offline': Boolean,
  'prefer-symlinked-executables': Boolean,
  'prefer-workspace-packages': Boolean,
  production: [null, true],
  'public-hoist-pattern': Array,
  'publish-branch': String,
  'recursive-install': Boolean,
  reporter: String,
  'resolution-mode': ['highest', 'time-based', 'lowest-direct'],
  'resolve-peers-from-workspace-root': Boolean,
  'aggregate-output': Boolean,
  'reporter-hide-prefix': Boolean,
  'save-peer': Boolean,
  'save-workspace-protocol': Boolean,
  'script-shell': String,
  'shamefully-flatten': Boolean,
  'shamefully-hoist': Boolean,
  'shared-workspace-lockfile': Boolean,
  'shell-emulator': Boolean,
  'side-effects-cache': Boolean,
  'side-effects-cache-readonly': Boolean,
  symlink: Boolean,
  sort: Boolean,
  'state-dir': String,
  'store-dir': String,
  stream: Boolean,
  'strict-store-pkg-content-check': Boolean,
  'strict-peer-dependencies': Boolean,
  'use-beta-cli': Boolean,
  'use-node-version': String,
  'use-running-store-server': Boolean,
  'use-store-server': Boolean,
  'use-stderr': Boolean,
  'verify-store-integrity': Boolean,
  'virtual-store-dir': String,
  'virtual-store-dir-max-length': Number,
  'peers-suffix-max-length': Number,
  'workspace-concurrency': Number,
  'workspace-packages': [String, Array],
  'workspace-root': Boolean,
  'test-pattern': [String, Array],
  'changed-files-ignore-pattern': [String, Array],
  'embed-readme': Boolean,
  'update-notifier': Boolean,
  'registry-supports-time-field': Boolean,
  'fail-if-no-match': Boolean,
}, npmTypes.types)
