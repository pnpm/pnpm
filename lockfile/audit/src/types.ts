export interface AuditVulnerabilityCounts {
  info: number
  low: number
  moderate: number
  high: number
  critical: number
}

export interface AuditResolution {
  id: number
  path: string
  dev: boolean
  optional: boolean
  bundled: boolean
}

export interface AuditAction {
  action: string
  module: string
  target: string
  isMajor: boolean
  resolves: AuditResolution[]
}

export interface AuditAdvisory {
  findings: [
    {
      version: string
      paths: string[]
      dev: boolean
      optional: boolean
      bundled: boolean
    }
  ]
  id: number
  created: string
  updated: string
  deleted?: boolean
  title: string
  found_by: {
    name: string
  }
  reported_by: {
    name: string
  }
  module_name: string
  cves: string[]
  vulnerable_versions: string
  patched_versions: string
  overview: string
  recommendation: string
  references: string
  access: string
  severity: string
  cwe: string
  metadata: {
    module_type: string
    exploitability: number
    affected_components: string
  }
  url: string
}

export interface AuditMetadata {
  vulnerabilities: AuditVulnerabilityCounts
  dependencies: number
  devDependencies: number
  optionalDependencies: number
  totalDependencies: number
}

export interface AuditReport {
  actions: AuditAction[]
  advisories: { [id: string]: AuditAdvisory }
  muted: unknown[]
  metadata: AuditMetadata
}

export interface AuditActionRecommendation {
  cmd: string
  isBreaking: boolean
  action: AuditAction
}
