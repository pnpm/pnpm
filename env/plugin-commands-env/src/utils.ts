import { promises as fs } from 'fs'
import path from 'path'

export const CURRENT_NODE_DIRNAME = 'nodejs_current'

export async function getNodeExecPathAndTargetDir (pnpmHomeDir: string): Promise<{ nodePath: string, nodeLink?: string }> {
  const nodePath = getNodeExecPathInBinDir(pnpmHomeDir)
  const nodeCurrentDirLink = path.join(pnpmHomeDir, CURRENT_NODE_DIRNAME)
  let nodeCurrentDir: string | undefined
  try {
    nodeCurrentDir = await fs.readlink(nodeCurrentDirLink)
  } catch (err) {
    nodeCurrentDir = undefined
  }
  return { nodePath, nodeLink: nodeCurrentDir ? getNodeExecPathInNodeDir(nodeCurrentDir) : undefined }
}

export function getNodeExecPathInBinDir (pnpmHomeDir: string): string {
  return path.resolve(pnpmHomeDir, process.platform === 'win32' ? 'node.exe' : 'node')
}

export function getNodeExecPathInNodeDir (nodeDir: string): string {
  return path.join(nodeDir, process.platform === 'win32' ? 'node.exe' : 'bin/node')
}
