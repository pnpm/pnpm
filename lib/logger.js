var supportsColor = require('supports-color')

module.exports = loggerType =>
  ((loggerType === 'pretty' && supportsColor)
    ? require('./logger/pretty')
    : require('./logger/simple'))()
