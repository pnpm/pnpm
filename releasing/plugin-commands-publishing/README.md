# @pnpm/plugin-commands-publishing

> The pack and publish commands of pnpm

[![npm version](https://img.shields.io/npm/v/@pnpm/plugin-commands-publishing.svg)](https://www.npmjs.com/package/@pnpm/plugin-commands-publishing)

## Installation

```sh
pnpm add @pnpm/plugin-commands-publishing
```

## License

MIT
