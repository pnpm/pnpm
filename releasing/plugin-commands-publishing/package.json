{
  "name": "@pnpm/plugin-commands-publishing",
  "version": "7.4.0",
  "description": "The pack and publish commands of pnpm",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib",
    "!*.map"
  ],
  "engines": {
    "node": ">=16.14"
  },
  "scripts": {
    "start": "tsc --watch",
    "lint": "eslint \"src/**/*.ts\" \"test/**/*.ts\"",
    "_test": "cross-env PNPM_REGISTRY_MOCK_PORT=7778 jest",
    "test": "pnpm run compile && pnpm run _test",
    "prepublishOnly": "pnpm run compile",
    "compile": "tsc --build && pnpm run lint --fix"
  },
  "repository": "https://github.com/pnpm/pnpm/blob/main/releasing/plugin-commands-publishing",
  "keywords": [
    "pnpm8",
    "pnpm",
    "pack",
    "publish"
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/pnpm/pnpm/issues"
  },
  "homepage": "https://github.com/pnpm/pnpm/blob/main/releasing/plugin-commands-publishing#readme",
  "devDependencies": {
    "@pnpm/filter-workspace-packages": "workspace:*",
    "@pnpm/plugin-commands-publishing": "workspace:*",
    "@pnpm/prepare": "workspace:*",
    "@pnpm/registry-mock": "3.16.0",
    "@types/cross-spawn": "^6.0.4",
    "@types/is-windows": "^1.0.1",
    "@types/proxyquire": "^1.3.30",
    "@types/ramda": "0.28.20",
    "@types/sinon": "^10.0.20",
    "@types/tar": "^6.1.7",
    "@types/tar-stream": "^2.2.3",
    "ci-info": "^3.9.0",
    "cross-spawn": "^7.0.3",
    "is-windows": "^1.0.2",
    "load-json-file": "^6.2.0",
    "path-exists": "^4.0.0",
    "tar": "^6.2.0",
    "write-yaml-file": "^5.0.0"
  },
  "dependencies": {
    "@pnpm/cli-utils": "workspace:*",
    "@pnpm/client": "workspace:*",
    "@pnpm/common-cli-options-help": "workspace:*",
    "@pnpm/config": "workspace:*",
    "@pnpm/error": "workspace:*",
    "@pnpm/exportable-manifest": "workspace:*",
    "@pnpm/fs.packlist": "workspace:*",
    "@pnpm/git-utils": "workspace:*",
    "@pnpm/lifecycle": "workspace:*",
    "@pnpm/package-bins": "workspace:*",
    "@pnpm/pick-registry-for-package": "workspace:*",
    "@pnpm/resolver-base": "workspace:*",
    "@pnpm/run-npm": "workspace:*",
    "@pnpm/sort-packages": "workspace:*",
    "@pnpm/types": "workspace:*",
    "@zkochan/rimraf": "^2.1.3",
    "enquirer": "^2.4.1",
    "execa": "npm:safe-execa@0.1.2",
    "fast-glob": "^3.3.1",
    "p-filter": "^2.1.0",
    "ramda": "npm:@pnpm/ramda@0.28.1",
    "realpath-missing": "^1.1.0",
    "render-help": "^1.0.3",
    "tar-stream": "^2.2.0",
    "tempy": "^1.0.1",
    "write-json-file": "^4.3.0"
  },
  "peerDependencies": {
    "@pnpm/logger": "^5.0.0"
  },
  "funding": "https://opencollective.com/pnpm",
  "exports": {
    ".": "./lib/index.js"
  }
}
