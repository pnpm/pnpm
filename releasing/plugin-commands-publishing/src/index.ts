import * as pack from './pack'
import * as publish from './publish'

export { pack, publish }
