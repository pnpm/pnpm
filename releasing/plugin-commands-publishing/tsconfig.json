{
  "extends": "@pnpm/tsconfig",
  "compilerOptions": {
    "outDir": "lib",
    "rootDir": "src"
  },
  "include": [
    "src/**/*.ts",
    "../../typings/**/*.d.ts"
  ],
  "references": [
    {
      "path": "../../cli/cli-utils"
    },
    {
      "path": "../../cli/common-cli-options-help"
    },
    {
      "path": "../../config/config"
    },
    {
      "path": "../../config/pick-registry-for-package"
    },
    {
      "path": "../../exec/lifecycle"
    },
    {
      "path": "../../exec/run-npm"
    },
    {
      "path": "../../packages/error"
    },
    {
      "path": "../../packages/git-utils"
    },
    {
      "path": "../../packages/types"
    },
    {
      "path": "../../pkg-manager/client"
    },
    {
      "path": "../../pkg-manager/package-bins"
    },
    {
      "path": "../../pkg-manifest/exportable-manifest"
    },
    {
      "path": "../../privatePackages/prepare"
    },
    {
      "path": "../../resolving/resolver-base"
    },
    {
      "path": "../../workspace/filter-workspace-packages"
    },
    {
      "path": "../../workspace/sort-packages"
    }
  ]
}
