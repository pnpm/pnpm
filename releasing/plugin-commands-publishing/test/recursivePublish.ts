import { promises as fs } from 'fs'
import path from 'path'
import { readProjects } from '@pnpm/filter-workspace-packages'
import { streamParser } from '@pnpm/logger'
import { publish } from '@pnpm/plugin-commands-publishing'
import { preparePackages } from '@pnpm/prepare'
import { REGISTRY_MOCK_PORT } from '@pnpm/registry-mock'
import { type ProjectManifest } from '@pnpm/types'
import execa from 'execa'
import crossSpawn from 'cross-spawn'
import loadJsonFile from 'load-json-file'
import { DEFAULT_OPTS } from './utils'

const CREDENTIALS = `\
registry=http://localhost:${REGISTRY_MOCK_PORT}/
//localhost:${REGISTRY_MOCK_PORT}/:username=username
//localhost:${REGISTRY_MOCK_PORT}/:_password=${Buffer.from('password').toString('base64')}
//localhost:${REGISTRY_MOCK_PORT}/:email=foo@bar.net`

test('recursive publish', async () => {
  // This suffix is added to the package name to avoid issue if Jest reruns the test
  const SUFFIX = Date.now()

  const pkg1 = {
    name: `@pnpmtest/test-recursive-publish-project-1-${SUFFIX}`,
    version: '1.0.0',

    dependencies: {
      'is-positive': '1.0.0',
    },
  }
  const pkg2 = {
    name: `@pnpmtest/test-recursive-publish-project-2-${SUFFIX}`,
    version: '1.0.0',

    dependencies: {
      'is-negative': '1.0.0',
    },
  }
  const projects = preparePackages([
    pkg1,
    pkg2,
    // This will not be published because is-positive@1.0.0 is in the registry
    {
      name: 'is-positive',
      version: '1.0.0',

      scripts: {
        prepublishOnly: 'exit 1',
      },
    },
    // This will not be published because it is a private package
    {
      name: 'i-am-private',
      version: '1.0.0',

      private: true,
      scripts: {
        prepublishOnly: 'exit 1',
      },
    },
    // Package with no name is skipped
    {
      location: 'no-name',
      package: {
        scripts: {
          prepublishOnly: 'exit 1',
        },
      },
    },
  ])

  await fs.writeFile('.npmrc', CREDENTIALS, 'utf8')

  await publish.handler({
    ...DEFAULT_OPTS,
    ...await readProjects(process.cwd(), []),
    dir: process.cwd(),
    dryRun: true,
    recursive: true,
  }, [])

  {
    const { status } = crossSpawn.sync('npm', ['view', pkg1.name, 'versions', '--registry', `http://localhost:${REGISTRY_MOCK_PORT}`, '--json'])
    expect(status).toBe(1)
  }
  {
    const { status } = crossSpawn.sync('npm', ['view', pkg2.name, 'versions', '--registry', `http://localhost:${REGISTRY_MOCK_PORT}`, '--json'])
    expect(status).toBe(1)
  }

  process.env.npm_config_userconfig = path.join('.npmrc')
  await publish.handler({
    ...DEFAULT_OPTS,
    ...await readProjects(process.cwd(), []),
    dir: process.cwd(),
    recursive: true,
  }, [])

  {
    const { stdout } = await execa('npm', ['view', pkg1.name, 'versions', '--registry', `http://localhost:${REGISTRY_MOCK_PORT}`, '--json'])
    expect(JSON.parse(stdout.toString())).toStrictEqual(pkg1.version)
  }
  {
    const { stdout } = await execa('npm', ['view', pkg2.name, 'versions', '--registry', `http://localhost:${REGISTRY_MOCK_PORT}`, '--json'])
    expect(JSON.parse(stdout.toString())).toStrictEqual(pkg2.version)
  }

  projects[pkg1.name].writePackageJson({ ...pkg1, version: '2.0.0' })

  await publish.handler({
    ...DEFAULT_OPTS,
    ...await readProjects(process.cwd(), []),
    dir: process.cwd(),
    recursive: true,
    tag: 'next',
  }, [])

  {
    const { stdout } = await execa('npm', ['dist-tag', 'ls', pkg1.name, '--registry', `http://localhost:${REGISTRY_MOCK_PORT}`])
    expect(stdout.toString()).toContain('next: 2.0.0')
  }
})

test('print info when no packages are published', async () => {
  preparePackages([
    // This will not be published because is-positive@1.0.0 is in the registry
    {
      name: 'is-positive',
      version: '1.0.0',

      scripts: {
        prepublishOnly: 'exit 1',
      },
    },
    // This will not be published because it is a private package
    {
      name: 'i-am-private',
      version: '1.0.0',

      private: true,
      scripts: {
        prepublishOnly: 'exit 1',
      },
    },
    // Package with no name is skipped
    {
      location: 'no-name',
      package: {
        scripts: {
          prepublishOnly: 'exit 1',
        },
      },
    },
  ])

  await fs.writeFile('.npmrc', CREDENTIALS, 'utf8')

  const reporter = jest.fn()
  streamParser.on('data', reporter)

  await publish.handler({
    ...DEFAULT_OPTS,
    ...await readProjects(process.cwd(), []),
    dir: process.cwd(),
    dryRun: true,
    recursive: true,
  }, [])

  streamParser.removeListener('data', reporter)
  expect(reporter).toBeCalledWith(expect.objectContaining({
    level: 'info',
    message: 'There are no new packages that should be published',
    name: 'pnpm',
    prefix: process.cwd(),
  }))
})

test('packages are released even if their current version is published, when force=true', async () => {
  preparePackages([
    // This version is already in the registry
    {
      name: 'is-positive',
      version: '3.1.0',

      scripts: {
        prepublishOnly: 'pnpm version major',
      },
    },
  ])

  await fs.writeFile('.npmrc', CREDENTIALS, 'utf8')

  await publish.handler({
    ...DEFAULT_OPTS,
    ...await readProjects(process.cwd(), []),
    force: true,
    dir: process.cwd(),
    dryRun: true,
    recursive: true,
  }, [])

  const manifest = await loadJsonFile<ProjectManifest>('is-positive/package.json')
  expect(manifest.version).toBe('4.0.0')
})

test('recursive publish writes publish summary', async () => {
  preparePackages([
    {
      name: '@pnpmtest/test-recursive-publish-project-3',
      version: '1.0.0',

      dependencies: {
        'is-positive': '1.0.0',
      },
    },
    {
      name: '@pnpmtest/test-recursive-publish-project-4',
      version: '1.0.0',

      dependencies: {
        'is-negative': '1.0.0',
      },
    },
    // This will not be published because is-positive@1.0.0 is in the registry
    {
      name: 'is-positive',
      version: '1.0.0',

      scripts: {
        prepublishOnly: 'exit 1',
      },
    },
    // This will not be published because it is a private package
    {
      name: 'i-am-private',
      version: '1.0.0',

      private: true,
      scripts: {
        prepublishOnly: 'exit 1',
      },
    },
    // Package with no name is skipped
    {
      location: 'no-name',
      package: {
        scripts: {
          prepublishOnly: 'exit 1',
        },
      },
    },
  ])

  await fs.writeFile('.npmrc', CREDENTIALS, 'utf8')

  process.env.npm_config_userconfig = path.join('.npmrc')
  await publish.handler({
    ...DEFAULT_OPTS,
    ...await readProjects(process.cwd(), []),
    dir: process.cwd(),
    recursive: true,
    reportSummary: true,
  }, [])

  {
    const publishSummary = await loadJsonFile('pnpm-publish-summary.json')
    expect(publishSummary).toMatchSnapshot()
    await fs.unlink('pnpm-publish-summary.json')
  }

  await publish.handler({
    ...DEFAULT_OPTS,
    ...await readProjects(process.cwd(), []),
    dir: process.cwd(),
    recursive: true,
    reportSummary: true,
  }, [])

  {
    const publishSummary = await loadJsonFile('pnpm-publish-summary.json')
    expect(publishSummary).toStrictEqual({
      publishedPackages: [],
    })
  }
})

test('when publish some package throws an error, exit code should be non-zero', async () => {
  preparePackages([
    {
      name: '@pnpmtest/test-recursive-publish-project-5',
      version: '1.0.0',
    },
    {
      name: '@pnpmtest/test-recursive-publish-project-6',
      version: '1.0.0',
    },
  ])

  // Throw ENEEDAUTH error when publish.
  await fs.writeFile('.npmrc', 'registry=https://__fake_npm_registry__.com', 'utf8')

  const result = await publish.handler({
    ...DEFAULT_OPTS,
    ...await readProjects(process.cwd(), []),
    dir: process.cwd(),
    recursive: true,
    force: true,
  }, [])

  expect(result?.exitCode).toBe(1)
})
