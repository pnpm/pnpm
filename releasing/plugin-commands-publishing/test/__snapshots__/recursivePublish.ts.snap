// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recursive publish writes publish summary 1`] = `
{
  "publishedPackages": [
    {
      "name": "@pnpmtest/test-recursive-publish-project-4",
      "version": "1.0.0",
    },
    {
      "name": "@pnpmtest/test-recursive-publish-project-3",
      "version": "1.0.0",
    },
  ],
}
`;
